src/docker.c

```diff
@@ -0,0 +1,1137 @@
+    }
+    get_interface_stats(interface_node, stats->network_stats->interfaces[i]);
+  }
+}
+
+// Parses the JSON response from the Docker STATS API and extracts metrics
+// for memory, disk, CPU and networking interfaces.
+static void extract_stats_from_response(char *response_buffer,
+    container_stats_t ** stats, const char *container_id) {
+  yajl_val node;
+  char errbuf[1024];
+  container_stats_t *ptr;
+  ptr = (container_stats_t *) calloc(1, sizeof(container_stats_t));
+  if (ptr == NULL) {
+    ERROR("docker: extract_stats_from_response: malloc failed!");
+    goto error;
+  }
+  node = yajl_tree_parse(response_buffer, errbuf, sizeof(errbuf));
+  if (node == NULL) {
+    if (strlen(errbuf)) {
+      ERROR("docker: parse_error: %s.\n", errbuf);
+    } else {
+      ERROR("docker: parse_error.\n");
+    }
+    goto error;
+  }
+  get_memory_stats(node, ptr);
+  get_blkio_stats(node, ptr);
+  get_cpu_stats(node, ptr, container_id);
+  get_network_stats(node, ptr);
+  extract_string(node, "name", &(ptr->name));
+  yajl_tree_free(node);
+  *(stats) = ptr;
+  return;
+
+error:
+  yajl_tree_free(node);
+  free_stats(ptr);
+}
+
+static container_stats_t *get_stats_for_container(const char *container_id,
+    const char *socket, const char *version) {
+  char *response_buffer = (char *) calloc(RESPONSE_BUFFER_SIZE, sizeof(char));
+  char *url = (char *) calloc(107, sizeof(char));
+  if (response_buffer == NULL || url == NULL) {
+    ERROR("docker: get_metrics_for_container: malloc failed.\n");
+  }
+  ssnprintf(url, 107, "http:/v%s/containers/%s/stats?stream=false", version,
+      container_id);
+  curl_get_json(response_buffer, RESPONSE_BUFFER_SIZE, url, socket);
+  container_stats_t *ptr;
+  extract_stats_from_response(response_buffer, &ptr, container_id);
+  sfree(response_buffer);
+  sfree(url);
+  return ptr;
+}
+
+//==============================================================================
+//==============================================================================
+//==============================================================================
+// Funcitonality to dispatch Collectd Value lists with metrics.
+//==============================================================================
+//==============================================================================
+//==============================================================================
+
+static void dispatch_value_list(const char *hostname, const char *plugin,
+  const char *plugin_instance, const char *type, const char *type_instance,
+  meta_data_t * md, size_t count, value_t value, ...) {
+  value_t values[count];
+  value_list_t vl = VALUE_LIST_INIT;
+  value_t val = value;
+  va_list ap;
+  va_start(ap, value);
+  for (int i = 0; i < count; i++) {
+    memcpy(&values[i], &val, sizeof(val));
+    val = va_arg(ap, value_t);
+  }
+  va_end(ap);
+
+  vl.values = values;
+  vl.values_len = count;
+  size_t plugin_name_len = 7 + strlen(plugin) + 1;
```

Comment:
Add a comment making it clear why you need the 7 - i.e., that you're adding "docker." to the plugin name.